cider/exec/nextgen/jitlib/llvmjit/LLVMJITModule.cpp

```diff
@@ -20,18 +20,52 @@
  */
 #include "exec/nextgen/jitlib/llvmjit/LLVMJITModule.h"
 
-#include <llvm/IR/Function.h>
+#include <llvm/Bitcode/BitcodeReader.h>
+#include <llvm/IR/Module.h>
+#include <llvm/Transforms/Utils/Cloning.h>
+
+#include <memory>
+#include <mutex>
 
 #include "exec/nextgen/jitlib/llvmjit/LLVMJITUtils.h"
 #include "util/Logger.h"
+#include "util/filesystem/cider_path.h"
 
 namespace cider::jitlib {
 
-LLVMJITModule::LLVMJITModule(const std::string& name)
+static llvm::MemoryBuffer* getExternalBuffer() {
+  static std::once_flag has_set_buffer;
+  static std::unique_ptr<llvm::MemoryBuffer> external_function_buffer;
+
+  std::call_once(has_set_buffer, [&]() {
+    auto root_path = cider::get_root_abs_path();
+    auto template_path = root_path + "/function/RuntimeFunctions.bc";
+    CHECK(boost::filesystem::exists(template_path));
+
+    auto buffer_or_error = llvm::MemoryBuffer::getFile(template_path);
+    CHECK(!buffer_or_error.getError()) << "bc_filename=" << template_path;
+    external_function_buffer = std::move(buffer_or_error.get());
+  });
+
+  return external_function_buffer.get();
+}
+
+LLVMJITModule::LLVMJITModule(const std::string& name, bool has_external_module)
     : context_(std::make_unique<llvm::LLVMContext>())
-    , module_(std::make_unique<llvm::Module>(name, *context_))
-    , engine_(nullptr) {
-  CHECK(context_);
+    , engine_(nullptr)
+    , has_external_module_(has_external_module) {
+  if (has_external_module_) {
+    auto expected_res =
+        llvm::parseBitcodeFile(getExternalBuffer()->getMemBufferRef(), *context_);
+    if (!expected_res) {
+      LOG(FATAL) << "LLVM IR ParseError: Something wrong when parsing bitcode.";
+    } else {
+      external_module_ = std::move(expected_res.get());
```

Comment:
why not assign it to `module_` directly? we don't need to save `external_module_`.
maybe we should compare the compile time using the whole runtime module and copy the used function. @bigPYJ1151 any thoughts?

Reply:
The parsed module contains the declarations and definitions of all functions, if assign directly, it will be very redundant in the IR. When directly copying a specific function, it is difficult to manually copy all of its dependent declarations. So at present, we will copy all declarations into the module, and copy the definition of the required function.

Reply:
> why not assign it to `module_` directly? we don't need to save `external_module_`. maybe we should compare the compile time using the whole runtime module and copy the used function. @bigPYJ1151 any thoughts?

Yes, many runtime function usage strategies need to explore and benchmarking. But now I think enable this feature is enough and we could find a better way later.

Reply:
LGTM